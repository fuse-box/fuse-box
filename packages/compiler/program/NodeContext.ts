export class NodeContext {
  public namespace?: string;
  constructor() {}
}

export function createNodeContext() {
  return new NodeContext();
}
