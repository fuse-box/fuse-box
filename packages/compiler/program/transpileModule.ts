import { ASTNode } from '../interfaces/AST';
import { ICompilerOptions } from '../interfaces/ICompilerOptions';
import { FastVisit, IVisit, IVisitorMod, TopLevelVisit } from '../Visitor/Visitor';

export type ITranformerCallback = (visit: IVisit) => IVisitorMod | void | undefined;
export type ITransformer =
  | ITranformerCallback
  | {
      onTopLevelTraverse?: ITranformerCallback;
      onEachNode?: ITranformerCallback;
    };

export type ITransformerList = Array<ITransformer>;
export interface IProgramProps {
  ast: ASTNode;
  globalContext?: any;
  namespace?: string;
  transformers: ITransformerList;
  compilerOptions?: ICompilerOptions;
}

export function transpileModule(props: IProgramProps) {
  const eachNodeTransformers: Array<ITranformerCallback> = [];
  const onTopLevelTransformers: Array<ITranformerCallback> = [];
  //console.log(JSON.stringify(props.ast, null, 2));
  for (const t of props.transformers) {
    if (typeof t === 'object') {
      if (t.onEachNode) eachNodeTransformers.push(t.onEachNode);
      if (t.onTopLevelTraverse) onTopLevelTransformers.push(t.onTopLevelTraverse);
    } else {
      eachNodeTransformers.push(t);
    }
  }

  props.globalContext.programProps = props;
  if (onTopLevelTransformers.length) {
    TopLevelVisit({
      ast: props.ast,
      globalContext: props.globalContext,
      fn: visit => {
        for (const onTopLevelCallback of onTopLevelTransformers) {
          let response = onTopLevelCallback(visit);
          if (response) return response;
        }
      },
    });
  }
  FastVisit({
    ast: props.ast,
    globalContext: props.globalContext,
    fn: visit => {
      for (const eachNodeCallback of eachNodeTransformers) {
        let response = eachNodeCallback(visit);
        if (response) return response;
      }
    },
  });
  if (props.globalContext.completeCallbacks.length) {
    for (const cb of props.globalContext.completeCallbacks) {
      cb();
    }
  }
}
