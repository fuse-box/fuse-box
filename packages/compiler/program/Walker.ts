import { NodeContext } from "./NodeContext";

export interface IVisitProps {
  parent?: any;
  prop?: string;
  idx?: number;
  context?: IASTContext;
}
export interface IASTContext {
  nodeContext?: NodeContext;
  locals?: Array<string>;
  exports?: Array<string>;
}

export interface IASTVisitor {
  node: any;
  props: IVisitProps;
  context?: IASTContext;
}
interface IASTWalkProps {
  withScope?: boolean;
  visit?: (props: IASTVisitor) => any;
}
export function fastWalk(ast: any, walker: IASTWalkProps) {
  function visit(node, props: IVisitProps, context) {
    if (walker.withScope) {
      if (node.context) {
        context = node.context;
        if (!node.context.locals) node.context.locals = [];
      }

      if (node.type === "VariableDeclaration") {
        if (node.declarations) {
          for (const decl of node.declarations) {
            if (
              decl.type === "VariableDeclarator" &&
              decl.id &&
              decl.id.type === "Identifier"
            ) {
              if (context === undefined) context = { locals: [] };
              context.locals.push(decl.id.name);
              // we need to check for the next item on the list (if we are in an array)
              if (props.idx && props.prop) {
                if (props.parent[props.prop][props.idx + 1]) {
                  props.parent[props.prop][props.idx + 1].context = context;
                }
              }
            }
          }
        }
      }

      if (node.type === "ExpressionStatement") {
        if (node.expression && node.expression.arguments) {
          node.expression.arguments.context = context;
        }
      }

      if (
        node.type === "FunctionDeclaration" ||
        node.type === "FunctionExpression"
      ) {
        if (node.id && node.id.type === "Identifier") {
          if (context === undefined) context = { locals: [] };
          context.locals.push(node.id.name);
          if (props.idx && props.prop) {
            if (props.parent[props.prop][props.idx + 1]) {
              props.parent[props.prop][props.idx + 1].context = context;
            }
          }
        }
        if (node.params) {
          for (const item of node.params) {
            if (item.type === "Identifier") {
              if (node.body.context === undefined) {
                // create context
                node.body.context = {
                  exports:
                    context && context.exports
                      ? context.exports.concat([])
                      : [],
                  // here we need to make a copy of the locals
                  locals: context ? context.locals.concat([]) : []
                };
              }
              node.body.context.locals.push(item.name);
            }
          }
        }
      }
    }

    const newNode = walker.visit({ node, props, context });
    if (newNode) {
      if (props.prop && props.parent) {
        if (props.idx) {
          props.parent[props.prop][props.idx] = newNode;
        } else {
          props.parent[props.prop] = newNode;
        }
      }
      node = newNode;
    }
    for (const prop in node) {
      if (prop[0] === "$") {
        continue;
      }
      const child = node[prop];
      if (child instanceof Array) {
        for (let i = 0; i < child.length; i++) {
          if (child && child[i] && child[i].type) {
            visit(child[i], { parent: node, prop, idx: i }, context);
          }
        }
      } else {
        if (child && child.type) visit(child, { parent: node, prop }, context);
      }
    }
  }

  visit(ast, {}, undefined);
}

export interface IVisit {
  node: any;
  parent?: any;
  prop?: string;
  idx?: number;
  context?: IASTContext;
}
export function FastVisit(ast, fn: (visit: IVisit) => void) {
  function _visit(
    node,
    props: { parent?; propName?; id? },
    context: IASTContext
  ) {}

  _visit(ast, {}, undefined);
}
