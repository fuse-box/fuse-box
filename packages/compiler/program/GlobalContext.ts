import { ASTNode } from '../interfaces/AST';
import { IProgramProps } from './transpileModule';

export interface GlobalContext {
  getNextIndex: () => number;
  exportAfterDeclaration?: {
    [key: string]: {
      target?: string;
    };
  };
  programProps?: IProgramProps;
  identifierReplacement: {
    [key: string]: {
      first?: string;
      second?: string;
      insertAfter?: ASTNode;
      inUse?: boolean;
    };
  };
  namespace?: string;
  completeCallbacks?: Array<() => void>;
}

export function createGlobalContext(userContext?: { [key: string]: any }): GlobalContext {
  let index = 1;
  let essentialContext = {
    completeCallbacks: [],
    getNextIndex: () => index++,
    identifierReplacement: {},
    namespace: 'exports',
  };
  if (userContext) {
    for (const key in userContext) {
      essentialContext[key] = userContext[key];
    }
  }

  return essentialContext;
}
