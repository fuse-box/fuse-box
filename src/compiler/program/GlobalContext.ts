import { ASTNode } from '../interfaces/AST';
import { IProgramProps } from './transpileModule';

export interface GlobalContext {
  getNextIndex: () => number;
  getNextSystemVariable: () => string;
  hoisted: { [key: string]: number };
  exportAfterDeclaration?: {
    [key: string]: {
      targets?: Array<string>;
    };
  };
  jsxFactory?: string;
  programProps?: IProgramProps;
  identifierReplacement: {
    [key: string]: {
      first?: string;
      second?: string;
      insertAfter?: ASTNode;
      inUse?: boolean;
    };
  };
  namespace?: string;
  completeCallbacks?: Array<() => void>;
}

const Letters = ['a', 'b', 'c', 'd', 'f', 'g', 'h', 'i'];

export function createGlobalContext(userContext?: { [key: string]: any }): GlobalContext {
  let VARIABLE_COUNTER = 0;
  let index = 1;
  let essentialContext = {
    completeCallbacks: [],
    hoisted: {},

    getNextIndex: () => index++,
    identifierReplacement: {},
    namespace: 'exports',
    getNextSystemVariable: () => {
      return `_${++VARIABLE_COUNTER}_`;
      //return `_${Letters[++VARIABLE_COUNTER]}`;
    },
  };
  if (userContext) {
    for (const key in userContext) {
      essentialContext[key] = userContext[key];
    }
  }

  return essentialContext;
}
