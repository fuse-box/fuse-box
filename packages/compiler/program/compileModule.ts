import * as buntis from "buntis";
import { generate } from "../generator/generator";
import { ASTNode } from "../interfaces/AST";
import { ICompilerOptions } from "../interfaces/ICompilerOptions";
import { ClassConstructorPropertyTransformer } from "../transformers/ClassConstructorPropertyTransformer";
import { ExportTransformer } from "../transformers/ExportTransformer";
import { GlobalContextTransformer } from "../transformers/GlobalContextTransformer";
import { ImportTransformer } from "../transformers/ImportTransformer";
import { InterfaceRemoverTransformer } from "../transformers/InterfaceRemoverTransformer";
import { JSXTransformer } from "../transformers/JSXTransformer";
import { NamespaceTransformer } from "../transformers/NameSpaceTransformer";
import { IVisit, IVisitorMod } from "../Visitor/Visitor";
import { createGlobalContext } from "./GlobalContext";
import { ITransformerList, transpileModule } from "./transpileModule";

export interface ICompileModuleProps {
  code: string;
  globalContext?: any;
  transformers?: Array<(globalContext) => (visit: IVisit) => IVisitorMod>;
  compilerOptions?: ICompilerOptions;
}

export function compileModule(props: ICompileModuleProps) {
  const ast = buntis.parseTSModule(props.code, {
    directives: true,
    jsx: true,
    next: true,
    loc: true,
    ts: true
  });

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    ClassConstructorPropertyTransformer(),
    JSXTransformer(),
    NamespaceTransformer(),
    InterfaceRemoverTransformer(),
    ImportTransformer(),
    ExportTransformer()
  ];
  transpileModule({
    ast: ast as ASTNode,
    compilerOptions: props.compilerOptions,
    globalContext: createGlobalContext(props.globalContext),
    transformers: defaultTransformers
  });
  //console.log(JSON.stringify(ast, null, 2));
  const res = generate(ast, {});

  return { code: res };
}
