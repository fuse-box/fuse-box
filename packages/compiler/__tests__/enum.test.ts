import { compileModule } from '../program/compileModule';

describe('Enums test', () => {
  it('should work with a simple enum', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          first,
          second,
        }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should work with a custom numner enum', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          first,
          second = 30,
        }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should compute binary', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          first,
          second = 1 << 2,
        }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should compute binary and do some math', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          first,
          second = 1 << 2 + 10 - 30 * (2 / 10),
        }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should compute binary and do some math with Math.function', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          first,
          second = 1 << 2 + 10 - 30 * Math.sin(2 / 3),
        }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should compute crazy enum and reference is in another enum', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          First = 1 << 1,
          Second = 1 << (2 + 10 - 30 * Math.sin(2 / 3)),
          Third = (((Second / 2) * 3) / First) * 2,
        }
    `,
    });
    expect(result.code).toMatchInlineSnapshot(`
      "var Stuff;
      (function (Stuff) {
        Stuff[Stuff[\\"First\\"] = 2] = \\"First\\"
        Stuff[Stuff[\\"Second\\"] = 67108864] = \\"Second\\"
        Stuff[Stuff[\\"Third\\"] = 100663296] = \\"Third\\"
      })(Stuff || (Stuff = {}))
      "
    `);
  });

  it('should respect local enum references if not able to compute', () => {
    const result = compileModule({
      code: `
        enum Stuff {
          First = 1 << 1,
          Second = Hey + 1 + First,
        }
    `,
    });
    expect(result.code).toMatchInlineSnapshot(`
            "var Stuff;
            (function (Stuff) {
              Stuff[Stuff[\\"First\\"] = 2] = \\"First\\"
              Stuff[Stuff[\\"Second\\"] = Hey + 1 + Stuff.First] = \\"Second\\"
            })(Stuff || (Stuff = {}))
            "
        `);
  });

  it.only('should have no problem exporting a enum', () => {
    const result = compileModule({
      code: `
        export enum Stuff {
          First = 1 << 1,
          Second = Hey + 1 + First,
        }
    `,
    });
    console.log(result.code);
  });
});
