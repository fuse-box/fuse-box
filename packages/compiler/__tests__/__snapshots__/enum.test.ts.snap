// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enums test should compute binary 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"first\\"] = 0] = \\"first\\"
  Stuff[Stuff[\\"second\\"] = 4] = \\"second\\"
})(Stuff || (Stuff = {}))
"
`;

exports[`Enums test should compute binary and do some math 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"first\\"] = 0] = \\"first\\"
  Stuff[Stuff[\\"second\\"] = 64] = \\"second\\"
})(Stuff || (Stuff = {}))
"
`;

exports[`Enums test should compute binary and do some math with Math.function 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"first\\"] = 0] = \\"first\\"
  Stuff[Stuff[\\"second\\"] = 67108864] = \\"second\\"
})(Stuff || (Stuff = {}))
"
`;

exports[`Enums test should compute crazy enum and reference is in another enum 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"First\\"] = 2] = \\"First\\"
  Stuff[Stuff[\\"Second\\"] = 67108864] = \\"Second\\"
  Stuff[Stuff[\\"Third\\"] = 100663296] = \\"Third\\"
})(Stuff || (Stuff = {}))
"
`;

exports[`Enums test should respect local enum references if not able to compute 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"First\\"] = 2] = \\"First\\"
  Stuff[Stuff[\\"Second\\"] = Hey + 1 + Stuff.First] = \\"Second\\"
})(Stuff || (Stuff = {}))
"
`;

exports[`Enums test should work with a custom numner enum 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"first\\"] = 0] = \\"first\\"
  Stuff[Stuff[\\"second\\"] = 30] = \\"second\\"
})(Stuff || (Stuff = {}))
"
`;

exports[`Enums test should work with a simple enum 1`] = `
"var Stuff;
(function (Stuff) {
  Stuff[Stuff[\\"first\\"] = 0] = \\"first\\"
  Stuff[Stuff[\\"second\\"] = 1] = \\"second\\"
})(Stuff || (Stuff = {}))
"
`;
